src/crypto/equihash.cpp

```diff
@@ -19,183 +19,210 @@
 {
-    for (int i = 0; i < len; i++)
-        hash[i] = a.hash[i];
+    eh_index i{t};
+    return (i << (ilen - 8)) | r;
 }
 
-StepRow& StepRow::operator=(const StepRow& a)
+template<size_t WIDTH>
+StepRow<WIDTH>::StepRow(unsigned int n, const eh_HashState& base_state, eh_index i)
 {
-    unsigned char* p = new unsigned char[a.len];
-    for (int i = 0; i < a.len; i++)
-        p[i] = a.hash[i];
-    delete[] hash;
-    hash = p;
-    len = a.len;
-    indices = a.indices;
-    return *this;
+    eh_HashState state;
+    state = base_state;
+    crypto_generichash_blake2b_update(&state, (unsigned char*) &i, sizeof(eh_index));
+    crypto_generichash_blake2b_final(&state, hash, n/8);
 }
 
-StepRow& StepRow::operator^=(const StepRow& a)
+template<size_t WIDTH> template<size_t W>
+StepRow<WIDTH>::StepRow(const StepRow<W>& a)
 {
-    if (a.len != len) {
-        throw std::invalid_argument("Hash length differs");
-    }
-    if (a.indices.size() != indices.size()) {
-        throw std::invalid_argument("Number of indices differs");
+    assert(W <= WIDTH);
+    std::copy(a.hash, a.hash+W, hash);
+}
+
+template<size_t WIDTH>
+FullStepRow<WIDTH>::FullStepRow(unsigned int n, const eh_HashState& base_state, eh_index i) :
+        StepRow<WIDTH> {n, base_state, i}
+{
+    EhIndexToArray(i, hash+(n/8));
+}
+
+template<size_t WIDTH> template<size_t W>
+FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, int trim) :
+        StepRow<WIDTH> {a}
+{
+    assert(len+lenIndices <= W);
+    assert(len-trim+(2*lenIndices) <= WIDTH);
+    for (int i = trim; i < len; i++)
+        hash[i-trim] = a.hash[i] ^ b.hash[i];
+    if (a.IndicesBefore(b, len)) {
+        std::copy(a.hash+len, a.hash+len+lenIndices, hash+len-trim);
+        std::copy(b.hash+len, b.hash+len+lenIndices, hash+len-trim+lenIndices);
+    } else {
+        std::copy(b.hash+len, b.hash+len+lenIndices, hash+len-trim);
+        std::copy(a.hash+len, a.hash+len+lenIndices, hash+len-trim+lenIndices);
     }
-    unsigned char* p = new unsigned char[len];
-    for (int i = 0; i < len; i++)
-        p[i] = hash[i] ^ a.hash[i];
-    delete[] hash;
-    hash = p;
-    indices.reserve(indices.size() + a.indices.size());
-    indices.insert(indices.end(), a.indices.begin(), a.indices.end());
-    return *this;
 }
 
-void StepRow::TrimHash(int l)
+template<size_t WIDTH>
+FullStepRow<WIDTH>& FullStepRow<WIDTH>::operator=(const FullStepRow<WIDTH>& a)
 {
-    unsigned char* p = new unsigned char[len-l];
-    for (int i = 0; i < len-l; i++)
-        p[i] = hash[i+l];
-    delete[] hash;
-    hash = p;
-    len -= l;
+    std::copy(a.hash, a.hash+WIDTH, hash);
+    return *this;
 }
 
-bool StepRow::IsZero()
+template<size_t WIDTH>
+bool StepRow<WIDTH>::IsZero(size_t len)
 {
     char res = 0;
     for (int i = 0; i < len; i++)
         res |= hash[i];
     return res == 0;
```

Comment:
I think this doesn't need to be constant-time, so it could exit early if ``hash[i] != 0``.